File: p14/main.py
"""
Longest Collatz Sequence solving
"""


def _count_sequence(n_start: int, lookup: dict = {}) -> int:
    """
    Returns the length of the collatz sequence when starting from
    n_start given a lookup. The function will also update the lookup
    once the value has been computed.
    """
    x, i = n_start, 0
    while x > 1:
        if x in lookup:
            lookup[n_start] = i + lookup[x]
            return i + lookup[x]
        if x % 2:
            x = 3 * x + 1
        else:
            x = x // 2
        i += 1

    lookup[n_start] = i
    return i


def solve(n_max=1_000_000) -> int:
    lookup = {}
    n_start, len_sequence = max(
        ((i, _count_sequence(i, lookup)) for i in range(n_max)), key=lambda x: x[1]
    )
    return n_start, len_sequence


if __name__ == "__main__":
    n_start, len_sequence = solve(n_max=1_000_000)
    print(
        "The starting number with the longest collatz sequence under "
        f"1'000'000 is {n_start} with a length of {len_sequence}"
    )

File: p14/src/main.rs
// Longest Collatz sequence solving

fn count_sequence(n_start: usize, lookup: &mut Vec<u64>) -> u64 {
    let mut x = n_start;
    let mut seq = Vec::new();
    let mut length = 0;

    while x > 1 {
        if let Some(&l) = lookup.get(x) {
            if l > 0 {
                length = l;
                break;
            }
        }

        seq.push(x);

        if x % 2 != 0 {
            x = 3 * x + 1;
        } else {
            x = x / 2;
        }
    }

    for &value in seq.iter().rev() {
        length += 1;
        if let Some(slot) = lookup.get_mut(value) {
            *slot = length;
        }
    }

    length
}

fn main() {
    let nmax: usize = 1_000_000;
    let mut lookup = vec![0u64; nmax];
    let mut cur: u64 = 0;

    for i in 1..nmax {
        let a = count_sequence(i, &mut lookup);
        if a > cur {
            cur = a;
        }
    }
    println!("{:?}", cur)
}
